### PR TITLE
Fix android letter spacing

### PR DESCRIPTION
**Pull request by tim-ford**
Motivation for this fix is to get "letterSpacing" property in styles to work on Android (for the most part - there are non working edge cases potentially - primary use case seems fine)

**Test plan (required)**
Tested using:
  letterSpacing: -6,
  fontSize:24,
  fontFamily:'ionicons',

Good Stars
goodstars

Bad Stars
badstars

### CODE REVIEW

**Comment by facebook-github-bot-1** [comment]
By analyzing the blame information on this pull request, we identified @MengjueW and @rigdern to be potential reviewers.

**Comment by facebook-github-bot-1** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-9** [comment]
@tim-ford updated the pull request.

**Comment by facebook-github-bot-4** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by nihgwu** [comment]
awesome

**Comment by facebook-github-bot-9** [comment]
@tim-ford updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@tim-ford updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9420/files/f3d931a2e557f659a4df980c257681da910bd1ed..5dab842c7ddc12f7fc52946bd77d615d579e3c66)

**Comment by munzur** [comment]
Any update on this?

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 1]
```diff
@@ -54,6 +56,19 @@ public void setText(ReactTextUpdate update) {
       (int) Math.ceil(update.getPaddingRight()),
       (int) Math.ceil(update.getPaddingBottom()));
 
+    float nextLetterSpacing = update.getLetterSpacing();
+    int fontSize = update.getFontSize();
+    if (!FloatUtil.floatsEqual(mLetterSpacing, nextLetterSpacing)) {
+      mLetterSpacing = nextLetterSpacing;
+      if(Float.isNaN(mLetterSpacing)) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: nit: space after `if`

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 2]
```diff
@@ -54,6 +56,19 @@ public void setText(ReactTextUpdate update) {
...
+        setLetterSpacing((float)0.0);
+      } else {
+        //calculate EM from proper font pixels
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: nit: space after `//`

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 3]
```diff
@@ -54,6 +56,19 @@ public void setText(ReactTextUpdate update) {
...
+        setLetterSpacing(1+(mLetterSpacing-PixelUtil.toDIPFromPixel(fontSize))/PixelUtil.toDIPFromPixel(fontSize));
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: Where does this formula come from - e.g. Android docs?

**Comment by mkonicek** [comment]
Thanks for adding this!

In the Test plan, could you please show screenshots of text that has no letterSpacing set (should be unchanged), and a few examples of text with a few different values of letterSpacing set?

This PR sets the letter spacing in `setText` - what's the reason of doing it there rather than in `setLetterSpacing`?

**Comment by facebook-github-bot-4** [comment]
@tim-ford do you have any updates for this pull request? It's been a while since the last update so wanted to check in and see if you've looked at the requested changes.

**Comment by belohlavek** [comment]
Will this be compatible with API 19+?
How long until it gets implemented?

**Comment by tim-ford** [comment]
the formula and fixes work -- the formula was researched on the web for PX to EM conversions since Android does letterSpacing in EM -- the reason this hasn't gone anywhere is: not sure why it fails the tests -- since everything works fine in a functional App

**Comment by tim-ford** [comment]
Also have updated fork to 35-stable -- not sure how to bring in those adjustments so the files can  no longer in conflice

**Comment by hubciorz** [comment]
Any news on this, guys?

**Comment by lacker** [comment]
@tim-ford Are the tests failing for you locally too, or do you think it's something going wrong with CircleCI? I am not sure what you mean exactly when you say "not sure why it fails the tests"

**Comment by tim-ford** [comment]
its been a couple of weeks since looking -- I have already released an application with this fix in it, the fix is fine -- updates have occurred since that will require some adjustments to code placement -- It seemed like an issue with CI at the time

**Comment by lacker** [comment]
The test errors do look related to spacing on Android so it at least isn't obvious to me that it's a flaky CI problem. But take a look and see if you agree with my take. Also fixing the merge conflicts should "kick" the CI to run again anyway.

**Comment by hramos** [comment]
Importing to see if our internal tests pass. It does seem like there are some conflicts that need to be resolved before we can merge this though.

**Comment by hramos** [comment]
@tim-ford can you resolve the conflicts?

**Comment by julesmoretti** [comment]
Would love this feature request to be resolved :)

**Comment by lacker** [comment]
It seems like this pull request has been abandoned. I'm going to close it but if you are still interested in working on this please feel free to reopen!

**Comment by jkomyno** [comment]
Why was this pull request abandoned?

**Comment by jerolimov** [comment]
Checkout #13199. I implemented letterSpacing based on this PR, but also solve that the layout was calculated correctly, which was missed here.
